lwotgtelem/gnmit/subscribe/subscribe.go

```diff
@@ -0,0 +1,596 @@
+			return status.Error(codes.Unauthenticated, "no authentication/authorization for requested operation")
+		}
+		c.acl = a
+	}
+	c.sr, err = stream.Recv()
+
+	switch {
+	case err == io.EOF:
+		return nil
+	case err != nil:
+		return err
+	case c.sr.GetSubscribe() == nil:
+		return status.Errorf(codes.InvalidArgument, "request must contain a subscription %#v", c.sr)
+	case c.sr.GetSubscribe().GetPrefix() == nil:
+		return status.Errorf(codes.InvalidArgument, "request must contain a prefix %#v", c.sr)
+	case c.sr.GetSubscribe().GetPrefix().GetTarget() == "":
+		return status.Error(codes.InvalidArgument, "missing target")
+	}
+
+	c.target = c.sr.GetSubscribe().GetPrefix().GetTarget()
+	if !s.c.HasTarget(c.target) {
+		return status.Errorf(codes.NotFound, "no such target: %q", c.target)
+	}
+	peer, _ := peer.FromContext(stream.Context())
+	mode := c.sr.GetSubscribe().Mode
+
+	log.Infof("peer: %v target: %q subscription: %s", peer.Addr, c.target, c.sr)
+	defer log.Infof("peer: %v target %q subscription: end: %q", peer.Addr, c.target, c.sr)
+
+	c.queue = coalesce.NewQueue()
+	defer c.queue.Close()
+
+	// reject single device subscription if not allowed by ACL
+	if c.target != "*" && !c.acl.Check(c.target) {
+		return status.Errorf(codes.PermissionDenied, "not authorized for target %q", c.target)
+	}
+	// This error channel is buffered to accept errors from all goroutines spawned
+	// for this RPC.  Only the first is ever read and returned causing the RPC to
+	// terminate.
+	errC := make(chan error, 3)
+	c.errC = errC
+
+	switch mode {
+	case pb.SubscriptionList_ONCE:
+		go func() {
+			s.processSubscription(&c)
+			c.queue.Close()
+		}()
+	case pb.SubscriptionList_POLL:
+		go s.processPollingSubscription(&c)
+	case pb.SubscriptionList_STREAM:
+		if c.sr.GetSubscribe().GetUpdatesOnly() {
+			if _, err := c.queue.Insert(syncMarker{}); err != nil {
+				return err
+			}
+		}
+		remove := addSubscription(s.m, c.sr.GetSubscribe(),
+			&matchClient{acl: c.acl, q: c.queue})
+		defer remove()
+		if !c.sr.GetSubscribe().GetUpdatesOnly() {
+			go s.processSubscription(&c)
+		}
+	default:
+		return status.Errorf(codes.InvalidArgument, "Subscription mode %v not recognized", mode)
+	}
+
+	go s.sendStreamingResults(&c)
+
+	return <-errC
+}
+
+////////////////////////// START -- Added Code from Fork of github.com/openconfig/gnmi ///////////////////////////////////
+
+// SubscribeLocal registers a set of paths to listen to. It returns a coalescing queue
+// object into which the updates will be sent, and a remove function that stops
+// the queue from being updated.
+func (s *Server) SubscribeLocal(target string, paths []*pb.Path, prefix *pb.Path) (*coalesce.Queue, func(), error) {
+	if !s.c.HasTarget(target) {
+		return nil, nil, status.Errorf(codes.NotFound, "no such target: %q", target)
+	}
+
+	log.Infof("target: %q SubscribeLocal start: %s", target, paths)
+	defer log.Infof("target %q SubscribeLocal start: end: %q", target, paths)
```

Comment:
nit: "start" here is extra? should be "target %q SubscribeLocal end: %q"?

Reply:
Good spot, thanks.